framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+     */
+    private String formatArrayComponentTypeAnnotation(ATypeElement e) {
+        StringBuilder result = new StringBuilder();
+        for (Map.Entry<InnerTypeLocation, ATypeElement> ite : e.innerTypes.entrySet()) {
+            InnerTypeLocation loc = ite.getKey();
+            AElement it = ite.getValue();
+            if (loc.location.contains(TypePathEntry.ARRAY)) {
+                result.append(formatAnnotations(it.tlAnnotationsHere));
+            }
+        }
+        return result.toString();
+    }
+
+    /**
+     * Formats an AField so that it can be printed in a stub. An AField represents a variable
+     * declaration. In practice, {@code aField} should represent either a field declaration or a
+     * formal parameter of a method, because stub files should not contain local variable
+     * declarations. It is the responsibility of each caller of this method to ensure that is true.
+     *
+     * @param aField the field to format
+     * @param fieldName the name to use for the declaration in the stub file. This doesn't matter
+     *     for parameters, but must be correct for fields.
+     * @param className the name of the enclosing class. This is only used for printing the type of
+     *     an explicit receiver parameter (i.e., a parameter named "this").
+     * @return a String suitable to print in a stub file
+     */
+    private String formatAField(AField aField, String fieldName, String className) {
+        StringBuilder result = new StringBuilder();
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else if (basetypes.containsKey(aField.type.description.toString())) {
+            basetype = basetypes.get(aField.type.description.toString()).toString();
+        } else {
+            throw new BugInCF(
+                    "SceneToStubWriter: could not find the base type for this variable declaration: "
+                            + fieldName);
+        }
+
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            result.append(formatArrayComponentTypeAnnotation(aField.type));
+            result.append(component);
+            result.append(" ");
+            basetype = "[]";
+        }
+        result.append(formatAnnotations(aField.type.tlAnnotationsHere));
+        result.append(basetype);
+        result.append(' ');
+        result.append(fieldName);
+        return result.toString();
+    }
+
+    /**
+     * Writes out an import statement for each annotation used in an {@link AScene}.
+     *
+     * <p>{@code DefCollector} is a facility in the Annotation File Utilities for determining which
+     * annotations are used in a given AScene. Here, we use that construct to write out the proper
+     * import statements into a stub file.
+     */
+    private static class ImportDefCollector extends DefCollector {
```

Comment:
The name `ImportDefCollector` is at odds with the documentation "*Writes*".  I suggest giving it a more informative name, like `ImportDefWriter`.